### PR TITLE
Add TypeScript typings. Fixes #2

### LINKED ISSUE #2: publish typings to index.d.ts

### PR DESCRIPTION
**Pull request by ericqweinstein**
I'm new enough to TypeScript that I'm _fairly_ sure this is the right direction, but not at all convinced. (I actually had a tougher time with the `export`/`import`s than the actual type annotations.)

Some thoughts:

1. It's not immediately clear to me if this is fine as-is or if I should introduce a [namespace or module](https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html). I tried exporting a namespace, but this caused namespace resolution issues in the generated JS, so I've left them out for now (it's super possible I was making a rookie mistake).
2. I'm not quite sure the best way to test this. I made a sample test `.ts` file (see below) that uses these annotations and got the correct breakages when I intentionally broke the type annotations, but I'm not sure this is enough; it might be better for me to create a skeleton TypeScript project and use `ought` to test it.
3. I took a stab at reasonable-looking options for the `tsconfig.json` based on [the documentation](https://www.typescriptlang.org/docs/handbook/compiler-options.html), but happy to make changes if these are too strict or if there's something I'm missing.

Tested with the following (version 3.3.3 of TS, 10.15.1 of Node):

## Happy Path: Passing Tests
```ts
import * as ought from "./index";

ought.equal(42, 42);
ought.notEqual(7, 11);
console.log(ought.config({ color: false }));
```
tests pass

## Sad Path: Type Mismatch
```ts
import * as ought from "./index";

ought.equal(42, '42');
ought.notEqual(7, 11);
console.log(ought.config({ color: false }));
```
type mismatch

## Sad Path: Failed Test
```ts
import * as ought from "./index";

ought.equal(42, 7);
ought.notEqual(7, 11);
console.log(ought.config({ color: false }));
```
failed test

CC @garybernhardt (no rush!)

### CODE REVIEW

**Comment by ericqweinstein** [comment]
(I'm also unsure if I should loosen the type annotation for `notEqual` to `notEqual(actual: any, expected: any, msg?: string): void`; I don't know if there's a use case for intentionally calling this on objects of different types.)

**Comment by garybernhardt** [comment]
My guess is that the simple `<T>` parameterization on `equal`/`notEqual` is right. It's probably best to start with the `<T>` and only change it if problems come up in practice.

It seems reasonable to wrap that simple test you wrote in a teenytest test, as long as it's not too much work. You can probably run that test easily with `ts-node` to avoid a build step. Maybe a parallel `test-types.ts` file alongside `test.js`?

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// Type definitions for ought 0.0.4
+// Project: https://github.com/testdouble/ought
+// TypeScript version: 3.3.3
+
+export function equal<T>(actual: T, expected: T, msg?: string): void;
+
+export function notEqual<T>(actual: T, expected: T, msg?: string): void;
+
+export function config(userConfig: object): object;
```
Comment (garybernhardt) [inline index.d.ts]: Maybe introduce a UserConfig interface as a type for userConfig?

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (garybernhardt) [inline tsconfig.json]: I'd remove `noImplicitAny`, `noImplicitThis`, and `strictNullChecks`, replacing them all with just `strict`, which implies those three and more. But also note that this tsconfig probably isn't even necessary unless there are tests committed to this repo; I don't think you'll need a tsconfig present for someone else to require your types.

File: index.d.ts (hunk 1 above)
Comment (ericqweinstein) [reply index.d.ts]: Oh, nice! I like that idea. 👍

**Comment by ericqweinstein** [comment]
Okay! These new tests should confirm that `ought` assertions work as intended in TypeScript (`test-types.ts`), as well as that the declarations are being used correctly (`test-type-error.ts`):

```sh
ericweinstein[~/work/ought] (feat/typings *) λ npm run test-types

> ought@0.0.4 test-types /Users/ericweinstein/work/ought
> bin/ts-test

Here's how (typed!) success looks and how fast it is
TAP version 13
1..1
ok 1 - "pass" - test #1 in `test-types.ts`
# Test run passed!
#   Passed: 1
#   Failed: 0
#   Total:  1

real	0m0.790s
user	0m0.866s
sys	0m0.074s
Here's how (typed!) failure looks and how fast it is
TAP version 13
1..1
{ maxDepth: 3,
  plugins: [],
  theme:
   { boolean: { open: '\u001b[33m', close: '\u001b[39m' },
     circular: '\u001b[90m[Circular]\u001b[39m',
     date: { invalid: '\u001b[31minvalid\u001b[39m', value: [Object] },
     diffGutters:
      { actual: '\u001b[31m-\u001b[39m ',
        expected: '\u001b[32m+\u001b[39m ',
        padding: '  ' },
     error: { ctor: [Object], name: [Object] },
     function: { name: [Object], stringTag: [Object] },
     global: { open: '\u001b[35m', close: '\u001b[39m' },
     item: { after: '\u001b[90m,\u001b[39m' },
     list:
      { openBracket: '\u001b[90m[\u001b[39m',
        closeBracket: '\u001b[90m]\u001b[39m' },
     mapEntry: { after: '\u001b[90m,\u001b[39m' },
     maxDepth: '\u001b[90m…\u001b[39m',
     null: { open: '\u001b[33m', close: '\u001b[39m' },
     number: { open: '\u001b[33m', close: '\u001b[39m' },
     object:
      { openBracket: '\u001b[90m{\u001b[39m',
        closeBracket: '\u001b[90m}\u001b[39m',
        ctor: [Object],
        stringTag: [Object],
        secondaryStringTag: [Object] },
     property:
      { after: '\u001b[90m,\u001b[39m',
        keyBracket: [Object],
        valueFallback: '\u001b[90m…\u001b[39m' },
     react:
      { functionType: '\u001b[90m⍟\u001b[39m',
        openTag: [Object],
        closeTag: [Object],
        tagName: [Object],
        attribute: [Object],
        child: [Object] },
     regexp: { source: [Object], flags: [Object] },
     stats: { separator: '\u001b[90m---\u001b[39m' },
     string:
      { open: '\u001b[34m',
        close: '\u001b[39m',
        line: [Object],
        multiline: [Object],
        controlPicture: [Object],
        diff: [Object] },
     symbol: { open: '\u001b[33m', close: '\u001b[39m' },
     typedArray: { bytes: [Object] },
     undefined: { open: '\u001b[33m', close: '\u001b[39m' } } }
not ok 1 - "fail" - test #1 in `test-types.ts`
  ---
  message: Expected actual to equal expected, but it did not:- 7
+ 11
  stacktrace: AssertionError [ERR_ASSERTION]: Expected actual to equal expected, but it did not:- 7
+ 11
    at new AssertionError (internal/errors.js:86:11)
    at throwError (/Users/ericweinstein/work/ought/lib/throw-error.js:5:9)
    at Object.equal (/Users/ericweinstein/work/ought/index.js:13:35)
    at Object.fail (/Users/ericweinstein/work/ought/test-types.ts:5:11)
    at /Users/ericweinstein/work/ought/node_modules/teenytest/lib/plugins/callbackify.js:14:21
    at runX (/Users/ericweinstein/work/ought/node_modules/teenytest/lib/plugins/wrap.js:22:7)
    at Object.userFunction [as wrap] (/Users/ericweinstein/work/ought/node_modules/teenytest/plugins/uncaught-exception.js:16:9)
    at callable (/Users/ericweinstein/work/ought/node_modules/teenytest/lib/plugins/wrap.js:29:24)
    at runX (/Users/ericweinstein/work/ought/node_modules/teenytest/lib/plugins/wrap.js:22:7)
    at Object.userFunction [as wrap] (/Users/ericweinstein/work/ought/node_modules/teenytest/plugins/results.js:10:9)
  ...
# Test run failed!
#   Passed: 0
#   Failed: 1
#   Total:  1
#
# Failures:
#
#   1 - "fail" - test #1 in `test-types.ts`
#
#     AssertionError [ERR_ASSERTION]: Expected actual to equal expected, but it did not:- 7
#     + 11
#         at new AssertionError (internal/errors.js:86:11)
#         at throwError (/Users/ericweinstein/work/ought/lib/throw-error.js:5:9)
#         at Object.equal (/Users/ericweinstein/work/ought/index.js:13:35)
#         at Object.fail (/Users/ericweinstein/work/ought/test-types.ts:5:11)
#         at /Users/ericweinstein/work/ought/node_modules/teenytest/lib/plugins/callbackify.js:14:21
#         at runX (/Users/ericweinstein/work/ought/node_modules/teenytest/lib/plugins/wrap.js:22:7)
#         at Object.userFunction [as wrap] (/Users/ericweinstein/work/ought/node_modules/teenytest/plugins/uncaught-exception.js:16:9)
#         at callable (/Users/ericweinstein/work/ought/node_modules/teenytest/lib/plugins/wrap.js:29:24)
#         at runX (/Users/ericweinstein/work/ought/node_modules/teenytest/lib/plugins/wrap.js:22:7)
#         at Object.userFunction [as wrap] (/Users/ericweinstein/work/ought/node_modules/teenytest/plugins/results.js:10:9)

real	0m0.894s
user	0m0.955s
sys	0m0.077s
Here's what type errors look like

/Users/ericweinstein/work/ought/node_modules/ts-node/src/index.ts:228
    return new TSError(diagnosticText, diagnosticCodes)
           ^
TSError: ⨯ Unable to compile TypeScript:
test-type-error.ts(5,21): error TS2345: Argument of type '"42"' is not assignable to parameter of type 'number'.
test-type-error.ts(9,18): error TS2345: Argument of type '{}' is not assignable to parameter of type 'UserConfig'.
  Property 'color' is missing in type '{}' but required in type 'UserConfig'.

    at createTSError (/Users/ericweinstein/work/ought/node_modules/ts-node/src/index.ts:228:12)
    at getOutput (/Users/ericweinstein/work/ought/node_modules/ts-node/src/index.ts:334:40)
    at Object.compile (/Users/ericweinstein/work/ought/node_modules/ts-node/src/index.ts:367:11)
    at Module.m._compile (/Users/ericweinstein/work/ought/node_modules/ts-node/src/index.ts:413:43)
    at Module._extensions..js (module.js:671:10)
    at Object.require.extensions.(anonymous function) [as .ts] (/Users/ericweinstein/work/ought/node_modules/ts-node/src/index.ts:416:12)
    at Module.load (module.js:573:32)
    at tryModuleLoad (module.js:513:12)
    at Function.Module._load (module.js:505:3)
    at Module.require (module.js:604:17)

real	0m0.824s
user	0m0.912s
sys	0m0.073s
```

File: tsconfig.json (hunk 2 above)
Comment (ericqweinstein) [reply tsconfig.json]: Since we're including `*.ts` tests, I think I'll go the `strict` route, but happy to modify as needed.

**Comment by searls** [comment]
Ok, if @garybernhardt gives a :+1: i'll merge and cut a release

**Comment by ericqweinstein** [comment]
Also @searls, if @garybernhardt gives the 👍, LMK if you'd like me to `rebase -i` down to a single commit

**Comment by garybernhardt** [comment]
I'm not sure what that big object in the test output is, but the changes look good to me!

**Comment by searls** [comment]
Blindly merged

**Comment by searls** [comment]
landed in 0.0.5

**Comment by ericqweinstein** [comment]
@garybernhardt I recently switched machines and things went sideways with iTerm2, so it's super possible this is local to me. (The same thing happens to me with the original tests, too.) Thanks so much for the review!

**Pull request merged**